This PR Introduces simple explaination for CycleGAN as part of Unit 5 : Generative Models (#158)
johko/computer-vision-course: chapters/en/Unit 5 - Generative Models/PRACTICAL APPLICATIONS & CHALLENGES/cycle_gan.mdx

# CycleGAN Introduction

> This section introduces CycleGAN, short for Cycle-Consistent Generative Adversarial Network, which is a framework designed for image-to-image translation tasks where paired examples are not available. Introduced by Zhu et al. in a [2017](https://arxiv.org/abs/1703.10593) paper, it represents a significant advancement in the field of computer vision and machine learning.

Reviewer: ```suggestion
This section introduces CycleGAN, short for *Cycle-Consistent Generative Adversarial Network*, which is a framework designed for image-to-image translation tasks where paired examples are not available. Introduced by Zhu et al. in a [2017](https://arxiv.org/abs/1703.10593) paper, it represents a significant advancement in the field of computer vision and machine learning.
```

In many image-to-image translation tasks, the goal is to learn a mapping between an input image and an output image. Traditional approaches often rely on large datasets of paired examples (e.g., photos and corresponding sketches). However, obtaining such paired datasets can be extremely challenging or even infeasible in many scenarios. This is where CycleGAN comes into play, as it is designed to work with unpaired datasets.

> ## What is Unpaired Image-to-Image translation

Reviewer: ```suggestion
## What Is Unpaired Image-to-Image Translation
```

Integration with Adversarial Loss: Cycle consistency loss is combined with adversarial loss, commonly used in GANs, to form a comprehensive loss function. This combined loss function is optimized simultaneously for both generators in CycleGAN.

**Least-Square Loss**:

It's a method that minimizes the sum of squared residuals. What that means is that it tries to find the best-fit line that has the smallest sum of squared distances between that line and all the points. In GANs, the **line** represents the label (real or fake), and the points represent the discriminator's predictions. The discriminator's loss function in Least Squares adversarial loss is formulated by calculating the squared distance between its predictions and the actual labels (real or fake) across multiple images. For the generator, the loss function is designed to make its fake outputs appear as real as possible, measured by how far these outputs deviate from the label of **real**. Having Least-Square Loss addresses vanishing gradients and mode collapse issues common in BCE loss.

**Indentity Loss**:

> Introduced in CycleGAN, identity loss is an optional loss term aimed at enhancing color preservation in generated images. It works by ensuring that an image input into a generator (e.g., a horse image into a zebra-to-horse generator) should ideally output the same image, as its already in the target style. The loss is calculated as the pixel distance between the real input and the output of the generator. A zero pixel distance (no change in the image) results in zero identity loss, which is the desired outcome.

Reviewer: ```suggestion
Introduced in CycleGAN, identity loss is an optional loss term aimed at enhancing color preservation in generated images. It works by ensuring that an image input into a generator (e.g., a horse image into a zebra-to-horse generator) should ideally output the same image, as it is already in the target style. The loss is calculated as the pixel distance between the real input and the output of the generator. A zero pixel distance (no change in the image) results in zero identity loss, which is the desired outcome.
```

For CycleGANs generators, this loss is added alongside adversarial and cycle consistency losses. It's adjusted using a lambda term (weighting factor).

![CycleGAN](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/cycleGAN1.jpg)

> This figure shows the combined GAN architecture functionality for both the **GAN**. These GANs are linked by cycle consistency, forming a cycle.

Reviewer: ```suggestion
This figure shows the combined GAN architecture functionality for both **GANs**. These GANs are linked by cycle consistency, forming a cycle.
```

> # GAN applications

Reviewer: Maybe a more representative heading?

> This sections introduces CycleGAN, short for Cycle-Consistent Generative Adversarial Network, is a framework designed for image-to-image translation tasks where paired examples are not available. 

Reviewer: ```suggestion
This section introduces CycleGAN, short for Cycle-Consistent Generative Adversarial Network, which is a framework designed for image-to-image translation tasks where paired examples are not available. 
```

Introduced by Zhu et al. in a [2017](https://arxiv.org/abs/1703.10593) paper, it represents a significant advancement in the field of computer vision and machine learning.

In many image-to-image translation tasks, the goal is to learn a mapping between an input image and an output image. 
Traditional approaches often rely on large datasets of paired examples (e.g., photos and corresponding sketches). 
However, obtaining such paired datasets can be extremely challenging or even infeasible in many scenarios. 
This is where CycleGAN comes into play, as it is designed to work with unpaired datasets.

## What is Unpaired Image-to-Image translation

{/* TODO : add unpaired_images.png and paired_images.png */}

> In many image translation scenarios, we encounter datasets that lack a direct, one-to-one correspondence between image pairs. This scenario is where unpaired image-to-image translation comes into play. 

Reviewer 2: ```suggestion
In many image translation scenarios, we encounter datasets lacking direct, one-to-one correspondence between image pairs. This scenario is where unpaired image-to-image translation comes into play. 
```

Here, instead of having matching pairs of images, you work with two distinct sets or "piles" of images, each representing a different style or domain, such as X and Y.
For instance, one pile might consist of realistic photographs, while the other could contain artworks by Monet, Cezanne, or other artists. Alternatively, the piles could represent different seasons, 
with one showcasing winter landscapes and the other summer scenes. Another example could be a collection of horse images in one pile and zebra images in the other, without any direct pairing between 

> the two. The objective in unpaired image-to-image translation is for the model to learn and extract the general stylistic elements from each pile and apply these learned styles to transform images 

Reviewer: This line is a continuation of the previous one, the line break is not really needed

> from one domain to another. This transformation often works both ways, allowing images from domain X to be translated into the style of domain Y, and vice versa. This approach is particularly valuable

Reviewer: same comment about line breaks here

when exact image pairs are unavailable or difficult to obtain.

> ## Main components of CycleGAN

Reviewer 2: ```suggestion
## Main Components of CycleGAN
```

> - Dual GAN Structure: 

Reviewer 2: ```suggestion
**Dual GAN Structure**
```

CycleGAN employs two GANs (Generative Adversarial Networks), one for translating from the first set to the second (e.g., zebra to horse) and another for the reverse process (horse to zebra). 
This dual structure ensures realism (via the adversarial process) and content preservation (via cycle consistency).
    - PatchGAN Discriminators: The discriminators used in CycleGAN are based on [PatchGAN](https://arxiv.org/pdf/1611.07004.pdf) architecture, which assesses patches of an image rather than the whole, contributing to more detailed and localized realism.

>     - Generator Architecture: The generators in CycleGAN draw from U-Net and DCGAN architectures, involving downsampling (encoding), upsampling (decoding), and convolutional layers with batch normalization and ReLU. 

Reviewer 2: This kinda assumes knowledge about U-Net and DCGAN, maybe you could give a link?

Author: Yes, I'll add the link.

These generators are enhanced with additional convolutional layers and skip connections, known as residual connections, which aid in learning identity functions and supporting deeper transformations.

- Cycle Consistency Loss:
Ensures that the style of an image can be changed (e.g., sad face to hugging face) and then reverted back to its original form (hugging face back to sad face) with minimal loss of detail or content.

> Implementation: Involves two stages of transformation:

Reviewer 2: ```suggestion
Implementation involves two stages of transformation.
```

**First Stage**: A sad face is transformed into a hugging face.
**Second Stage**: This hugging face is then converted back into a sad face.

The model aims for the final image (reverted sad face) to closely resemble the original sad face. This is achieved by minimizing the pixel difference between these two images, which is added to the model's loss function.

The cycle consistency is applied in both directions: sad face to hugging face to sad face, and hugging face to sad face to hugging face.
The loss for each cycle is calculated by summing the pixel differences and is then incorporated into the overall generator loss.

> Integration with Adversarial Loss:

Reviewer 2: ```suggestion

**Integration with Adversarial Loss**
```

Cycle consistency loss is combined with adversarial loss, commonly used in GANs, to form a comprehensive loss function.
This combined loss function is optimized simultaneously for both generators in CycleGAN.

> - Least-Square Loss

Reviewer 2: ```suggestion
**Least-Square Loss**
```

> It's a method that minimizes the sum of squared residuals. What that means is that it tries to find the best fit line that has the smallest sum of squared distances between that line and all the points.

Reviewer: ```suggestion
It's a method that minimizes the sum of squared residuals. What that means is that it tries to find the best-fit line that has the smallest sum of squared distances between that line and all the points.
```

In GANs, the **line** represents the label (real or fake), and the points represent the discriminator's predictions. The discriminator's loss function in Least Squares adversarial loss is formulated by 
calculating the squared distance between its predictions and the actual labels (real or fake) across multiple images. For the generator, the loss function is designed to make its fake outputs appear 
as real as possible, measured by how far these outputs deviate from the label of **real**.
Having Least-Square Loss addresses vanishing gradients and mode collapse issues common in BCE loss.

> -Indentity Loss 

Reviewer 2: ```suggestion
**Indentity Loss**
```

Introduced in CycleGAN, identity loss is an optional loss term aimed at enhancing color preservation in generated images. 
It works by ensuring that an image input into a generator (e.g., a horse image into a zebra-to-horse generator) should ideally output the same image, as its already in the target style.
The loss is calculated as the pixel distance between the real input and the output of the generator. A zero pixel distance (no change in the image) results in zero identity loss, which is the desired outcome.
For CycleGANs generators, this loss is added alongside adversarial and cycle consistency losses. It's adjusted using a lambda term (weighting factor).

> ![CycleGAN](https://cdn-uploads.huggingface.co/production/uploads/633c3172ec4e4abf307b7dc6/rcUT5Sa1jRPlrUvrr_iyq.jpeg)

Reviewer 2: You can give the link like this: https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/CycleGAN2.jpg also if this is made by you it looks very nice!

Author: Thank you so much @merveenoyan
Yes, I made it :) It was not merged before that's why added temporary like that.

> This Figure shows the combined GAN architecture functionality for both the **GAN**. These GANs are linked by cycle consistency, forming a cycle.

Reviewer: ```suggestion
This figure shows the combined GAN architecture functionality for both the **GAN**. These GANs are linked by cycle consistency, forming a cycle.
```

> This Figure shows the combined GAN architecture functionality for both the **GAN**. These GANs are linked by cycle consistency, forming a cycle.

Reviewer 2: ```suggestion
This Figure shows the combined GAN architecture functionality for both the **GAN**. These GANs are linked by cycle consistency, forming a cycle.
```
both the GAN and what? did you mean both of the GANs?

Author: did you mean both of the GANs? ---- Yess. I'll write it properly.
